Source/WebKit/Shared/Cocoa/CoreIPCLocale.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if PLATFORM(COCOA)
+
+#import <Foundation/Foundation.h>
+#import <wtf/text/WTFString.h>
+
+OBJC_CLASS NSLocale;
+
+namespace WebKit {
+
+class CoreIPCLocale {
+public:
+    static bool isValidIdentifier(const String& identifier)
+    {
+        if ([[NSLocale availableLocaleIdentifiers] containsObject:identifier])
+            return true;
+        if (replacement(identifier))
+            return true;
+        return false;
+    }
+
+    CoreIPCLocale(NSLocale *locale)
+        : m_identifier([locale localeIdentifier])
+    {
+    }
+
+    CoreIPCLocale(String&& identifier)
+        : m_identifier([[NSLocale currentLocale] localeIdentifier])
+    {
+        if ([[NSLocale availableLocaleIdentifiers] containsObject:identifier])
+            m_identifier = identifier;
+        else if (auto fixedLocale = replacement(identifier))
+            m_identifier = *fixedLocale;
+    }
+
+    RetainPtr<id> toID() const
+    {
+        return [[NSLocale alloc] initWithLocaleIdentifier:(NSString *)m_identifier ];
+    }
+
+    String identfier() const
+    {
+        return m_identifier;
+    }
+
+private:
+    static std::optional<String> replacement(String identifier);
```

Comment:
This needs a better name that indicates what it does.  It should probably also take a const String&